src/main/drivers/dshot_bitbang_decode.c

```diff
@@ -0,0 +1,290 @@
+#include "drivers/dshot.h"
+#include "drivers/dshot_bitbang_decode.h"
+
+#define MIN_VALID_BBSAMPLES ((21 - 2) * 3)
+#define MAX_VALID_BBSAMPLES ((21 + 2) * 3)
+
+// setting this define in dshot.h allows the cli command dshot_telemetry_info to
+// display the received telemetry data in raw form which helps identify
+// the root cause of packet decoding issues.
+
+#ifdef DEBUG_BBDECODE
+uint16_t bbBuffer[134];
+#endif
+
+
+/* Bit band SRAM definitions */
+#define BITBAND_SRAM_REF   0x20000000
+#define BITBAND_SRAM_BASE  0x22000000
+#define BITBAND_SRAM(a,b) ((BITBAND_SRAM_BASE + ((a-BITBAND_SRAM_REF)<<5) + (b<<2)))  // Convert SRAM address
+
+typedef struct bitBandWord_s {
+    uint8_t value;
+    uint8_t junk[63];
+} bitBandWord_t;
+
+#ifdef DEBUG_BBDECODE
+uint32_t sequence[MAX_GCR_EDGES];
+int sequenceIndex = 0;
+#endif
+
+
+static uint32_t decode_bb_value(uint32_t value, uint16_t buffer[], uint32_t count, uint32_t bit)
+{
+#ifndef DEBUG_BBDECODE
+    UNUSED(buffer);
+    UNUSED(count);
+    UNUSED(bit);
+#endif
+#define iv 0xffffffff
+    // First bit is start bit so discard it.
+    value &= 0xfffff;
+    static const uint32_t decode[32] = {
+        iv, iv, iv, iv, iv, iv, iv, iv, iv, 9, 10, 11, iv, 13, 14, 15,
+        iv, iv, 2, 3, iv, 5, 6, 7, iv, 0, 8, 1, iv, 4, 12, iv };
+    
+    uint32_t decodedValue = decode[value & 0x1f];
+    decodedValue |= decode[(value >> 5) & 0x1f] << 4;
+    decodedValue |= decode[(value >> 10) & 0x1f] << 8;
+    decodedValue |= decode[(value >> 15) & 0x1f] << 12;
+    
+    uint32_t csum = decodedValue;
+    csum = csum ^ (csum >> 8); // xor bytes
+    csum = csum ^ (csum >> 4); // xor nibbles
+    
+    if ((csum & 0xf) != 0xf || decodedValue > 0xffff) {
+#ifdef DEBUG_BBDECODE
+        memcpy(dshotTelemetryState.inputBuffer, sequence, sizeof(sequence));
+        for (unsigned i = 0; i < count; i++) {
+            bbBuffer[i] = !!(buffer[i] & (1 << bit));
+        }
+#endif
+        value = BB_INVALID;
+    } else {
+        value = decodedValue >> 4;
+        
+        if (value == 0x0fff) {
+            return 0;
+        }
+        // Convert value to 16 bit from the GCR telemetry format (eeem mmmm mmmm)
+        value = (value & 0x000001ff) << ((value & 0xfffffe00) >> 9);
+        if (!value) {
+            return BB_INVALID;
+        }
+        // Convert period to erpm * 100
+        value = (1000000 * 60 / 100 + value / 2) / value;
+    }
+    return value;
+}
+
+
+uint32_t decode_bb_bitband( uint16_t buffer[], uint32_t count, uint32_t bit)
+{
+#ifdef DEBUG_BBDECODE
+    memset(sequence, 0, sizeof(sequence));
+    sequenceIndex = 0;
+#endif
+    uint32_t value = 0;
+
+    bitBandWord_t* p = (bitBandWord_t*)BITBAND_SRAM((uint32_t)buffer, bit);
+    bitBandWord_t* b = p;
+    bitBandWord_t* endP = p + (count - MIN_VALID_BBSAMPLES);
+
+    // Eliminate leading high signal level by looking for first zero bit in data stream.
+    // Manual loop unrolling and branch hinting to produce faster code.
```

Comment:
(BTW: did you reviewed generated code? unrolling may not be worth it)

Reply:
Yeah it's definitely worth it. I was hoping for 2 instructions per sample but unfortunately compiler generates 3 (doesn't manage to use CJZ, CJNZ). Considering there may be 9 consecutive samples of the same level not having to do the p increase, the p < endP comparison and the branch for each bit does make a big difference.

Reply:
(Using uint32_t may help a bit ... )

Reply:
Nah - it's just ldrw vs ldrb or something similar. arm does widening in the same instruction, no cycle cost.

Reply:
On the other hand when searching the first zero we could do all bits on one port together and  look at the not bit banded buffer using 32 bit to check two data points in one step. That would help a bit. Someday ...

Reply:
Yes, but each ESC may reply with different delay, so it will handle only common part ...

Reply:
Which is the main part since we don't detect the response or predict when it'll arrive. But icing on the cake.